### ISSUE TITLE
can't see toots on people's profiles or replies to many toots from self-hosted instance

### ISSUE THREAD

**Issue by chaosmonk1** [mastodon/mastodon#14017]
### Expected behaviour

When I view someone's profile locally, I should be able to see their public toots.

If a toot appears in my federated feed, clicking on the toot should display public replies to it.

### Actual behaviour

For many users, I can only see one or two toots when I view their profile on my local instance, but can see many toots when I view their profile on the remote instance.

Some toots don't appear to have any replies when viewed from my instance, but when I click on the date to view the toot on the remote instance, I see that there are replies.

### Steps to reproduce the problem

If necessary I can link to my instance and to examples of profiles missing toots and toots missing replies.

### Specifications

I am running Mastodon 3.1.4.

**Comment by SSPJ** [comment]
Same issue (?), also 3.1.4. The `mastodon-web.service` logs don't appear to have anything useful. This is what I get when I visit a profile page of a remote user I am following (line breaks for ease of reading):

```
... method=GET path=/api/v1/accounts/9/identity_proofs format=html \
controller=Api::V1::Accounts::IdentityProofsController action=index status=200 \
duration=11.66 view=1.45 db=5.84
... method=GET path=/api/v1/accounts/9/statuses format=html  \
controller=Api::V1::Accounts::StatusesController action=index status=200 \
duration=28.12 view=0.83 db=7.34
... method=GET path=/api/v1/accounts/9/statuses format=html \
controller=Api::V1::Accounts::StatusesController action=index status=200 \
duration=65.41 view=19.38 db=20.46
```

Some of the users I am following do not show up in the federated timeline at all, while others have just pinned toots, or (seemingly) random toots. None have the full public timeline that I can see when I visit the remote instance.

**Comment by ThibG** [comment]
This is expected, your instance only shows toots it knows about, and it won't pull toots from remote users. Some people suggest changing that (#1745) but it's not obvious how to do that correctly and efficiently.

We used to have the UI tell, on remote profiles, that they might be incomplete or outdated, but for some reason that part of the UI has been removed.

**Comment by chaosmonk1** [comment]
Aghhh, sorry for dragging this out, and thanks for your patience.

So to clarify:

* Say I am on instance `example.com`, and the only person that I or anyone else on `example.com` follows is `alice@mastodon.social`.  Alice follows `bob@mastodon.online`.
* When Bob publishes a toot,  `mastodon.online` fowards the toot to `mastodon.social`, where it shows up in Alice's federated timeline.
* She boosts or replies to Bob's toot, at which point `mastodon.social` fowards Bob's toot and Alice's reply or boost to `example.com`, where it appears in my federated timeline.
* Then someone else replies to Bob's toot, and `mastodon.online` forwards that reply to `mastodon.social` where Alice sees it.

* The current behavior at this point is that `mastodon.social` does **not** forward the new reply to `example.com`, which is why I don't see it. I behavior I was suggesting is that `mastodon.social` instead **would** forward the new reply to `example.com`.  In other words, currently `mastodon.social` says, "Okay, Alice interacted with Bob's toot and someone on `example.com` follows Alice, so I had better forward Bob's toot and Alice's boost or reply to `example.com` so that they can see it", and my suggestion was that `mastodon.social` would also say, "Additionally, from now on when I become aware of a new reply to Bob's toot, I will forward that reply to `example.com` as well".

* The problem is that forwarding these replies would be too expensive for `mastodon.social` and `example.com`?

**Comment by ThibG** [comment]
>     * When Bob publishes a toot,  `mastodon.online` fowards the toot to `mastodon.social`, where it shows up in Alice's federated timeline.

I would say “sends” rather than “forwards”, as the toot originates from `mastodon.online` and not something else, but yes.

>     * She boosts or replies to Bob's toot, at which point `mastodon.social` fowards Bob's toot and Alice's reply or boost to `example.com`, where it appears in my federated timeline.

Technically, `mastodon.social` doesn't forward the toot to `example.com`, just sends the boost/reply, which references the toot on `mastodon.online`, and upon receiving the boost/reply, `example.com` resolves that reference by fetching the original toot.

>     * Then someone else replies to Bob's toot, and `mastodon.online` forwards that reply to `mastodon.social` where Alice sees it.

Yes.

>     * The current behavior at this point is that `mastodon.social` does **not** forward the new reply to `example.com`, which is why I don't see it.

Yes.

> In other words, currently `mastodon.social` says, "Okay, Alice interacted with Bob's toot and someone on `example.com` follows Alice, so I had better forward Bob's toot and Alice's boost or reply to `example.com` so that they can see it",

not really, Mastodon just sends the reply/boost to Alice's followers and does not perform any extra step here, it's the receiving end which discovers a new reference and tries fetching it

> and my suggestion was that `mastodon.social` would also say, "Additionally, from now on when I become aware of a new reply to Bob's toot, I will forward that reply to `example.com` as well".
> 
>     * The problem is that forwarding these replies would be too expensive for `mastodon.social` and `example.com`?

Yes, that last suggestion would involve, for each received toot, check all the ancestry and for each of those ancestor toots, check who they have been boosted by, etc., which is expensive. Additionally, it would mean that multiple servers would ultimately forward you the same messages, which is a waste. And with all of that, it wouldn't even guarantee the forwarding would always work: for instance, consider what happens if Alice stops following Bob, or she unboosts his toot, or someone gets blocked along the way, etc. You'd stop getting replies.

**Comment by chaosmonk1** [comment]
Okay, that makes sense.  Thanks for explaining.

I think the issue is that the UI creates an expectation of different behavior from what actually happens.  I was hoping that the behavior could be changed to match the UI, but since it sounds like this is not feasible, I think the next best thing would be to change the UI to match the behavior.

Currently, the simplest possible interaction with a toot (clicking on it) displays some replies, and there is nothing to indicate that other replies are missing and can only be viewed on the remote instance.  This creates an expection of a very simple behavior (when I click on a toot, I can see the replies).  It is not obvious that there are additional replies that can only be seen by clicking on the date of the toot, and when the user eventually discovers this behavior, it is unclear why this is the case for some replies and not others.

In situations where Mastodon presents potentially incomplete information (replies to a toot, toots on a remote user's profile) it would be helpful if the UI indicated ideally
* what subset of the information is being shown
* what subset of the information is not being shown
* how to access the latter subset

or if it is too difficult to concisely summarize what information is/isn't shown, at least
* that some information may be missing
* how to access it

A simple form that this could take is a "view all toots" button on a remote user's profile, which links to the user's profile on the remote instance, and a "view all replies" on a toot by a remote user, which links to the toot on the remote instance.

**Comment by kuba-orlik** [comment]
@ThibG 

> Yes, that last suggestion would involve, for each received toot, check all the ancestry and for each of those ancestor toots, check who they have been boosted by, etc., which is expensive

I don't know much about Masto infrastructure, but in this scenario, can't the replies be just "pulled" from the original instance? I mean, with some coding duct tape they could even be scraped from the web interface of the orig instance

**Comment by ThibG** [comment]
Yes, Mastodon could theoretically pull replies (some of them at least) from the original instance (with some added machinery). But that expensive. And the question is when and how often? If you don't do it every single time someone clicks on the toot, you are likely to miss some replies. And even then, doing it at that moment will often be “too late” unless we add an asynchronous API for fetching replies and accept that the UI would suddenly have a bunch of new posts seconds after displaying the initial post.

**Comment by kuba-orlik** [comment]
Hm, I see. It's a shame though, because it makes it harder for small instances to view the entire discussion. 

Maybe an indicator saying "some replies might be missing" and a "pull more replies" button would be sufficcient?

**Comment by Krinkle** [comment]
I'm encountering the same issue on polyglot.city (on masto.host), and observe it both when viewing a profile after searching for it, and after having followed them. I tried this in a web app (Pinafore), and in two mobile apps (Amaroq, and Tootle).

I was nearly sure this was a bug in one of the apps, until I realized that other apps suffer the same issue. I then thought it was a problem with Masto Host, but someone pointed me here, that it is in fact a more general issue that just happens to affect smaller instances the most.

| Web | Amaroq app | Tootle app
|--|--|--
|![capture-web-pinafore](https://user-images.githubusercontent.com/156867/95908454-03cd3380-0d95-11eb-9e5d-072280eb4a41.png) |![Amaroq](https://user-images.githubusercontent.com/156867/95908719-53abfa80-0d95-11eb-9d3f-c32264db0f40.PNG) | ![Tootle](https://user-images.githubusercontent.com/156867/95908722-54449100-0d95-11eb-8edb-aa321fd0deea.PNG)

I can sort of understand and explain to people around me, that replies and like/boost counts can be incomplete when looking at a specific toot, if that toot is from someone else, outside your local group.

However, I am unable to explain to people why it is that Mastodon lets them paste/search for any account, and click on accounts of boosted toots or comments, but then cannot "just" browse those account's profile and public toots. We let them click on it and show recent meta data, but then the profile contains only a handful of random/old toots. I think this might be negatively affecting adoption and growth among communities as it means accounts look unappealing or abandoned.

I don't know much about the inner workings of Mastodon. I can totally see how it might be infeasible to federate these ahead of time, and undesirable to locally federate these on-demand when the user isn't followed. But, that suggest to me a few things:

* Perhaps we should not consult the local database at all then when browsing a profile unless it is followed. Instead, use an upstream API query, possibly cached locally in some way. (Either naively, by query; or normalized into an efficient internal structure separate from the timeline data, with stronger cache expiry or something).
* Once a new account is followed, we either import "everything" (unlikely), or import just one toot and remember in some way the timestamp of their last public toot and/or when they were followed by non-zero accounts, and for any incomplete query crossing over that date, do the same as for unfollowed accounts.

I also note that we don't really promote web links to the public profile, and on mobile apps these links are often omitted from the interface entirely. I suppose an alternate solution could be to encourage apps to short-circuit links to non-followed accounts and have them point to public account URLs always (thus for apps, switching to a browser and removing the ability to easily follow them without resorting to separately logging in and/or tech-savvy gestures like copying URLs into clipboards), where some apps might choose to instead query upstream APIs directly from the client-side (which might be undesirable in terms of privacy and performance/caching/latency). If this direction is preferred, it might make sense to deprecate or otherwise emit a warning to apps when querying profile data for accounts that are not followed an known to be sparse/outdated.

See also:
* "proxy pagination" ref https://github.com/tootsuite/mastodon/issues/14023#issuecomment-689026176

**Comment by chaosmonk1** [comment]
Ok, I've spent a few months keeping an open mind and trying to get used to this behavior, but I still think that this is a serious problem, at least for small instances.  The way I've adapted is to just always open toots and profiles via the remote instance in a new tab.  I don't feel comfortable replying to a toot until I have done this, because otherwise I don't know what context I might be missing.  At that point, Mastodon might as well just proxy all of the replies from the remote instance in the first place.  The added latency would surely be less than the time it takes to open up the remote instance in a new tab and load its web interface plus all of the replies.

**Comment by Krinkle** [comment]
I too open profiles in a new tab via the browser (outside any app or web app I use), and then copy-paste the urls of the latest 10 toots into the local search field on my home instance so that they get imported and show up for the next person.

Whatever reservations we might have about how importing affects or doesn't affect the federated timeline, it seems like this is something Mastodon could faccilitate to do more easily and is effectively already happening. Both by hand as I do, as well as for pinned toots. Perhaps in through the same handling as for pinned toots that are updated/imported, it could e.g. also fetch the last 10 toots on-demand and import them if not already present.

**Comment by mind-overflow** [comment]
hey, two years have passed and this issue is still open, is there any update?

**Comment by vorletzter** [comment]
I finally searched for this and stumbled on this issue.

i understand why this is handled the way it is handled, but as more and more instances keep popping up, i really would like the option to follow (or at least view) a thread from my instance directly instead of making my instance follow all the participating users and instances. This seems like a waste of resources in the long run.

**Comment by heartlessglasses** [comment]
Is this still an issue?

**Comment by kuba-orlik** [comment]
Yes, very much so

**Comment by alisonw** [comment]
Having run up my own instance over the weekend I'd say that this is, though undeniably mostly for small instances where there aren't already users connected to the particular instance you want updates from. Using a 'major' instance and my 'local' instance following the same accounts elsewhere (in both cases) gets very different visibility of past toots. A way to force this, maybe to the last X days or Y toots, would be very beneficial. The short-term machine excess load isn't necessarily a big issue where it is a personal / manually-initiated task.

**Comment by ConiferAri** [comment]
One relatively simple way to mitigate this, would be to offer the option to open a federated post on its home server directly.

Currently you have to use "copy link to post" then paste into a browser window.
Basically, add a new button which does this for you.

**Comment by johncronan** [comment]
People are talking about this a lot. I heard about the issue before jumping in, but figured "it can't be that bad" and went to go see for myself.

Yikes. I'm baffled, because the protocol is just broken, in this spot? From a user point of view, it is a fatal flaw.

Mitigating it by using a larger instance is a bandaid, doesn't fully solve the problem, and is contrary to the project's ideals.

I self host, so for me it's very bad. At first I just thought, nobody replies to anything in this place!

Really nice job on the Mastodon server implementation, though! Design flaws in the protocol are certainly not your fault. How can this be fixed?

**Comment by cl4rknova** [comment]
second this too. Relativly new to mastodon and activitypub in general I run un the same trap. In my case with self hosting. Never used twitter, but in germany I see a lot people migrating from there and for now it seems like they haven't recocgnized those issues, because most of them are living in a bubble. But they will. And then they will leave mastodon, because they will be disappointed. Also, there is a kind of "false labeling" for mastodon, if you are new. I for my self encouraged to host my self. Because the bannerat the registration form says it would easily possible to host by yourself. And because I try to host most of my stuff, I did. No hint about the downsides. 
IMHO there should be high priority in fixing those issues, or giving a clear statement about this from the beginning and encourage people in a very positiv wording, to try this "paradigm shift" in comparison to other platforms and to see those circumstances as a chance. For whatever.The way this is done right now will produce lots of discontent for sure in the future.

**Comment by cl4rknova** [comment]
... but not only selfhosting a small or single user instance is the problem. Have two accounts on bigger instances. And also there stuff is not consistent. More or less they are in sync, but I am missing posts there from time to time, although I have seen posts from the same account there before. Sometimes there seem to be "skips" in consistency.

**Comment by johncronan** [comment]
Having read through the thread more carefully, there's a lot of great detail on the issues that factor into this!

I don't think it's that the protocol is broken in this regard, anymore. But there are certainly a number of steps required before the problem can get down to a negligible size.

I'd like to look into it further, what it would take to have replies from users a particular remote doesn't follow periodically forwarded to them if they have reason to receive, as @chaosmonk1 suggests. That part's the biggest contributor from what I've seen so far, hosting an instance with just me on it so far. Idea would be, after some time at least, what's shown is a complete representation (barring any particularly weird circumstances).

**Comment by m-ghaoui** [comment]
If only there was a peer-to-peer protocol for sharing information 🤔 ...

**Comment by LeonardoPaccianiMori** [comment]
@m-ghaoui that might be a solution. However, I think there might be a couple of simpler ones that don't require creating more files and occupying more server space per user.
Mastodon already has an RSS feed for each user, so each account already has a .rss file associated to it. Wouldn't it be simpler to pull/recreate the old toots from that?
Or alternatively, pull only the last n toots (with let's say n=5) and then add a button at the end of the page saying "Load more" that pulls the next n toots from that user, and so on. This way the amount of requests to servers would be reduced.

Wouldn't this be more feasible? Or am I missing something?

**Comment by johncronan** [comment]
@LeonardoPaccianiMori, the solution proposed in #18150 is a really well thought out form of this type of mainly client-side fix.

I think the challenges of doing this fully on the server side are pretty formidable, so it would be good to support what's going on over there in the meantime.

**Comment by LeonardoPaccianiMori** [comment]
@johncronan I see. I'll catch up with that conversation and try to contribute as much as I can.

**Comment by lrossouw** [comment]
Agreed that this is important as it's quite easy to comment in a large thread without knowing.  I had replied to a toot the other day thinking I'm the first to reply but I essentially barged in on a big conversation that had already been going on. If my server is going to show me a toot it should also best effort attempt to fetch the whole context of the toot.  Probably finding the root toot :) and finding all replies to that.

**Comment by danger89** [comment]
I also miss this feature actually. I run my own server. I only see the reply on the toot that I follow (plus the toot itself), but not all the other replies. For that I need to click on "Open original page" and view the toot on the other server. Which is not ideal.

Hopefully this can become a setting of my server. Of course, the person who I follow should maybe on-top of all the rest of the replies.

**Comment by robsonsobral** [comment]
Please, [take a look](https://github.com/mastodon/mastodon/issues/17213#issuecomment-1356870891)

**Comment by larryedwards** [comment]
What isn't clear to me is this thread and related ones is where the data for a thread is actually stored. Is is stored only on the server of thread's original post (and if so what happens if that server is closed down?); or is it stored on multiple servers that have made their own individual interactions with some part of the thread?

It seems to me that there are too many mailmen involved in reconstructing a thread. If all data is on the thread's originating server (TOS), all servers which get reply/boost/favorite clicks would report to the TOS. The TOS would then mail a notice to all servers that need notification (having someone(s) who follows the OP or the thread), and would also report the counts of replies/boosts/favorites when requested by other servers, and maintain the thread's data tree. 

That would avoid the info loss that users are experiencing, and if I understand the current system correctly it would be more efficient and less costly to run and maintain.

**Comment by indirect** [comment]
FYI, at least one version of this problem might be because of a configuration issue. I booted my instance once before setting the LOCAL_DOMAIN env var, and that resulted in the "representative" account for the instance storing the username `localhost:3000`. It seems some servers will not respond to requests from a server with the username "localhost:3000", and so my instance was not able to fetch those posts.

The symptoms I saw were:
- some posts were missing from the federated timeline
- using search on the full url of some posts from other instances would return no results

I was able to fix the problem for my instance by running this command:

```
bin/rails run 'Account.representative.update(username: ENV["LOCAL_DOMAIN"])'
```

After that, searching for posts by the full URL started working, and posts from other instances were visible.

**Comment by alexisart** [comment]
Yeah. I'd like to be able to fully sync the list of emojis and public toots from other servers to my local instance

**Comment by trwnh** [comment]
#34 and #9409 cover this, i think

**Comment by runevision** [comment]
> #34 and #9409 cover this, i think

In the issue "Backfill statuses from remote accounts when first subscribed" - does "subscribe" mean to follow a profile? Or simply to view it? (There may be technical/implementation terminology I'm not familiar with.) Because the issue with posts not being shown is very much an issue when just viewing a profile, whether following that profile or not. Often, what content is available on a profile informs whether to follow it or not in the first place.

So if "subscribe" mean follow, then #34 and #9409 do not cover the present issue.

**Comment by vladh** [comment]
#32615 closed this issue — but should it have?

My understanding is that #32615 makes it so that, when you view a remote post, you can see all its replies.

This issue (#14017) does mention this problem. But it also mentions a _different_ problem: when you view the profile of a remote user you don't follow, you can't see all their posts.

My understanding is that this latter problem still exists, and is not fixed by #32615. I've also experienced this problem on my own self-hosted server.

So I wonder if this issue (#14017) should be re-opened.

Sorry if I misunderstood anything!
